app/src/main/java/com/sdp13epfl2021/projmag/activities/ProjectInformationActivity.kt

```diff
@@ -108,30 +113,65 @@ class ProjectInformationActivity : AppCompatActivity() {
         }
     }
 
+    private fun onApplyClick(
+        applyButton: Button,
+        candidatureDatabase: CandidatureDatabase,
+        projectId: ProjectId,
+    ) {
+        if (alreadyApplied) {
+            candidatureDatabase.removeCandidature(
+                projectId,
+                userId!!,
+                {
+                    showToast(getString(R.string.success), Toast.LENGTH_SHORT)
+                    alreadyApplied = !alreadyApplied
+                    setApplyButtonText(applyButton, alreadyApplied)
+                },
+                { showToast(getString(R.string.failure), Toast.LENGTH_SHORT) }
+            )
+        } else {
+            candidatureDatabase.pushCandidature(
+                projectId,
+                userId!!,
+                Candidature.State.Waiting,
+                {
+                    showToast(getString(R.string.success), Toast.LENGTH_SHORT)
+                    alreadyApplied = !alreadyApplied
+                    setApplyButtonText(applyButton, alreadyApplied)
+                },
+                { showToast(getString(R.string.failure), Toast.LENGTH_SHORT) }
+            )
+        }
+    }
+
     private fun setUpApplyButton(applyButton: Button) {
         val projectId = projectVar.id
-        val userDataDatabase = Utils.getInstance(this).userdataDatabase
-        var alreadyApplied = false
+        val utils = Utils.getInstance(this)
+        val userdataDatabase = utils.userdataDatabase
+        val candidatureDatabase = utils.candidatureDatabase
         setApplyButtonText(applyButton, null)
```

Comment:
There is already a util that is stored inside the class. You could also store the userdataDatabase and candidature database inside the class instead of invoking them in the function like this